### ISSUE TITLE
Hook annotations ignored on `PriorityClass` resource

### ISSUE THREAD

**Issue by consideRatio** [kubernetes/helm#4277]
## TL;DR
1. Should Helm expedite loading of PriorityClass resources? Perhaps... But it can be done through a helm hook, so... 
2. I made a typo, spent many hours, made a sloooow facepalm.

![picard-facepalm](https://user-images.githubusercontent.com/3837114/42126690-e848c0bc-7c8b-11e8-8d89-6fc89191717e.jpg)

---

## Issue 1 - Prioritized loading of PriorityClasses?

By adding a global [`PriorityClass`](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#priorityclass), all new pods without a `priorityClassName` will get some declared (pod-) `priority`.

```yaml
apiVersion: scheduling.k8s.io/v1alpha1
kind: PriorityClass
metadata:
  name: global-priority
value: 10
globalDefault: true
description: "A global priority for all new pods created after this is loaded."
```

By including this `PriorityClass` in my chart, I hoped that all new object created would be assigned the global priority, but they wasn't until they were restarted manually after the `helm install` command. So I figure the `PriorityClass` was not loaded with expedition by Helm.

#### Expedite loading of PriorityClasses
Does it make sense to expedite loading of PriorityClasses within Helm?

---

## Issue 2 - Helm ignores the hook annotation on PriorityClasses
I attempted to work around this by forcing Helm to load it first using the prioritization possibilities using weights in Helm hooks. The lower weight value, the earlier you will be installed. It turns out this did not work out of the box.

```yaml
apiVersion: scheduling.k8s.io/v1alpha1
kind: PriorityClass
metadata:
  name: global-priority
  annotation:
    # This global PriorityClass must be added first!
    helm.sh/hook: pre-install,pre-upgrade
    helm.sh/hook-delete-policy: before-hook-creation
    helm.sh/hook-weight: "-100"
value: 10
globalDefault: true
description: "A global priority for all new pods created after this is loaded."
```

I figured it might be because I also may require some time to pass in between this creation of the PriorityClass and the other pods. To accomplish this I created a Job with a pause container to run until I manually killed it after the PriorityClass. This still did not work! The PriorityClass will only be added after the blocking job is completed even though the PriorityClass is marked by the helm hook annotation to be loaded into k8s first!

So now I don't know whats going on any more... I've got this working with a different type of resource like a DaemonSet being loaded before a blocking Job.

#### My research

This issue has not arisen when I do the exact same thing with a DaemonSet in another context. Is Helm ignores hook annotations on PriorityClasses simply handling the resource as if it was not a hook?

Yes - Helm is ignoring the hook annotation on this PriorityClass! I tested this by removing the `helm.sh/hook-delete-policy: before-hook-creation` annotation and making one fresh chart install followed by an update to the resource and helm treated it as an versioned update rather than a hook resource to be created! If it did, it would had complained about the resource already existing but it didn't! Instead of saying a resource was already having that name, I got the complaint that the `priority` field could not be updated.

To conclude I raise the question - *why does Helm not honor hook annotations on PriorityClass resources?*

#### Note to self
It could be nice to use a helm hook with a `helm.sh/hook-delete-policy: before-hook-creation` annotation on PriorityClass resources to avoid using the `--force` flag on `helm upgrade` while making changes to it.

## Environment

```
$ helm version --debug
Client: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}
Kubernetes: &version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.4-gke.2", GitCommit:"eb2e43842aaa21d6f0bb65d6adf5a84bbdc62eaf", GitTreeState:"clean", BuildDate:"2018-06-15T21:48:39Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
Server: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}
```

## Reproduction output
```
$ helm install <the stuff below...>

$ kubectl get priorityclass
No resources found.

$ kubectl delete job post-default-priority-pause
job.batch "post-default-priority-pause" deleted

$ kubectl get priorityclass
NAME              AGE
global-priority   5s
```

## Reproduction resources

```yaml
apiVersion: scheduling.k8s.io/v1alpha1
kind: PriorityClass
metadata:
  name: global-priority
  annotation:
    # This global PriorityClass must be added first.
    helm.sh/hook: pre-install,pre-upgrade
    helm.sh/hook-delete-policy: before-hook-creation
    helm.sh/hook-weight: "-100"
value: 10
globalDefault: true
description: "A default priority higher than the placeholder priority."
---
apiVersion: batch/v1
kind: Job
metadata:
  name: post-default-priority-pause
  annotations:
    helm.sh/hook: pre-install
    helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded,hook-failed
    helm.sh/hook-weight: "-90"
spec:
  template:
    metadata:
      labels:
        key1: value1
    spec:
      restartPolicy: Never
      containers:
        - name: pause
          image: gcr.io/google_containers/pause:3.1
```

**Comment by jascott1** [comment]
Hi @consideRatio 

What version of k8s and Helm are you using and on what platform?

**Comment by consideRatio** [comment]
Heyy @jascott1 !

Helm 2.9.1 (client/server)
Kubernetes 1.10.4-gke.2
Googles Kubernetes engine
My cluster has alpha features enabled

**Comment by consideRatio** [comment]
## UPDATE
I've figured the underlying issue is that the annotations of the PriorityClass resource is stripped somewhere along the way when loaded into kubernetes. The question now is where and why?

I'm working on a v1.10.4 alpha enabled GKE cluster. They support PriorityClasseses, and my kubectl can add annotations as well.

If I load the following PriorityClass with an annotation, it is stripped silently.

```yaml
apiVersion: scheduling.k8s.io/v1alpha1
kind: PriorityClass
metadata:
  name: default-priority
  annotation:
    key1: value1
value: 10
globalDefault: true
description: "A default priority higher than the placeholder priority."
```

But if I then later make a `helm upgrade` where I've updated the `value1` to `newvalue`, I get the following error.

> Error: UPGRADE FAILED: failed to create patch: unable to find api field in struct ObjectMeta for the json field "annotation"

@jascott1 does that ring any bells for you? I'm not confident on whats going on here, but I figure I'm on the right track now.

**Comment by consideRatio** [comment]
### I forgot the `s` on `annotations` and had silent failures...
![picard-facepalm](https://user-images.githubusercontent.com/3837114/42126429-f575f69a-7c88-11e8-972a-284a51ba0ac2.jpg)

---

```yaml
metadata:
  annotation:
    key1: value1
```
